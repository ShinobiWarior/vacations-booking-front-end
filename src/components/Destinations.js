import React from 'react';
import { useSelector } from 'react-redux';
import Flickity from 'react-flickity-component';
import { Card } from 'react-bootstrap';
import '../styles/Flickity.css';
import '../styles/Destinations.css';
import { Link } from 'react-router-dom';

const Destinations = () => {
  const data = useSelector((state) => state.destinations.destinations);
  let state = [];
  if (data) state = data.sort(() => 0.5 - Math.random());
  const isLoading = useSelector((state) => state.destinations.loading);

  return (
    <>
      { isLoading ? (
        <img className="w-100" src="rotate-pulsating-loading-animation.webp" alt="spinner" />
      ) : (
        <div className="container-fluid align-items-center d-flex">
          <div className="align-middle w-100">
            <h4 className="d-flex justify-content-center mb-5 text-info fw-bold">Latest Destinations</h4>
            <h5 className="d-flex justify-content-center mb-5 text-info fw-bold">
              {' '}
              Please select Favourite Destination
              {' '}
            </h5>
            <Flickity
              className="carousel" // default ''
              elementType="div" // default 'div'
            // options="flickityOptions" // takes flickity options {}
              disableImagesLoaded={false} // default false
              reloadOnUpdate // default false
              static
            >
              {state.map((data) => (
                <Card key={data.id}>
                  <Link to={`/destinations/${data.id}`} className="dest-link">
                    <Card.Img className="rounded-circle" variant="top" src={data.image_url} alt="Card  cap" />
                    <Card.Body>
                      <Card.Title>
                        <h4 className="d-flex justify-content-center ">
                          {data.name}
                        </h4>
                      </Card.Title>
                      <Card.Text>
                        <small>Lorem dbkr tgbp rg3wtwposv wrbweuhsrog sgsruehorg efskg</small>
                      </Card.Text>
                    </Card.Body>
                    <Card.Footer>
                      <div className="d-flex justify-content-center">
                        <i className="fa fa-facebook-official fa-lg" />
                        <i className="fa fa-twitter fa-lg" />
                        <i className="fa fa-linkedin-square fa-lg" />
                      </div>
                    </Card.Footer>
                  </Link>
                </Card>
              ))}
            </Flickity>
          </div>
        </div>
      )}
    </>
  );
};

export default Destinations;
