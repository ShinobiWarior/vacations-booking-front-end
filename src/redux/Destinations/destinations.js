import getDestinations from '../../helpers/getDestinations';
import destroyDestination from '../../helpers/destroyDestination';

// Actions
const LOAD = 'vacations-booking-front-end/destinations/LOAD';
const DELETE = 'vacations-booking-front-end/destinations/DELETE';
const IS_LOADING = 'vacations-booking-front-end/destinations/IS_LOADING';
const initialState = [];

// Action creators
const deleteDestination = (payload) => ({
  type: DELETE,
  payload,
});

const deleteThunkDestination = (destinationId) => async (dispatch) => {
  const response = await destroyDestination();
  if (response.status === 'Success!') {
    dispatch(deleteDestination(destinationId));
  }
};

const loadDestinations = (payload) => ({
  type: LOAD,
  payload,
});

const loadDestinationsThunk = () => async (dispatch) => {
  const data = await getDestinations();
  if (data) {
    dispatch(loadDestinations(data));
  }
};

const destIsLoading = () => ({
  type: IS_LOADING,
});

// reducer
const destinationsReducer = (state = initialState, action) => {
  const { type, payload } = action;
  switch (type) {
    case LOAD:
      return {
        ...state,
        destinations: payload,
        loading: false,
      };
    case DELETE:
      return {
        ...state,
        ...state.destinations.filter((destination) => destination.id !== action.payload),
        loading: false,
      };
    case IS_LOADING:
      return {
        ...state,
        loading: true,
      };

    default:
      return state;
  }
};

export {
  destinationsReducer, loadDestinationsThunk, deleteThunkDestination, destIsLoading,
};
