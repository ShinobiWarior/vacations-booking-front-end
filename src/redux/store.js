import { createStore, combineReducers, applyMiddleware } from 'redux';
import logger from 'redux-logger';
import thunk from 'redux-thunk';
import { persistStore, persistReducer } from 'redux-persist';
import storage from 'redux-persist/lib/storage'; // defaults to localStorage for web
import { reservationReducer } from './Reservations/reservation';
import authReducer from './Auth/auth';
import { destinationsReducer } from './Destinations/destinations';

const persistConfig = {
  key: 'root',
  storage,
};

const reducer = combineReducers({
  reservations: reservationReducer,
  destinations: destinationsReducer,
  user: authReducer,
});
const persistedReducer = persistReducer(persistConfig, reducer);

const store = createStore(persistedReducer, applyMiddleware(logger, thunk));
export const persistor = persistStore(store);

export default store;
