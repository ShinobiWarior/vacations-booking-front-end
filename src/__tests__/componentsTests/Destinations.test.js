import React from 'react';
import renderer, { create } from 'react-test-renderer';
import { Provider } from 'react-redux';
import Destinations from '../../components/Destinations';
import store from '../../redux/store';

describe('Destinations', () => {
  it('renders correctly', () => {
    const tree = renderer
      .create(<Provider store={store}><Destinations /></Provider>)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
  test('the className of the component includes container', () => {
    const { root } = create(<Provider store={store}><Destinations /></Provider>);
    const element = root.findByType('div');
    expect(element.props.className.includes('container')).toBe(true);
  });
  test('the  component to have h4 element', () => {
    const { root } = create(<Provider store={store}><Destinations /></Provider>);
    const element = root.findByType('h4');
    expect(element.type).toBe('h4');
  });
  test('the  component to have h5 element with right content', () => {
    const { root } = create(<Provider store={store}><Destinations /></Provider>);
    const element = root.findByType('h5');
    expect(element.children.includes('Please select Favourite Destination')).toBe(true);
  });
});
