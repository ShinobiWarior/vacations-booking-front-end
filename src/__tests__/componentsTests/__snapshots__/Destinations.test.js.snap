// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Destinations renders correctly 1`] = `
<div
  className="container"
>
  <div
    className="align-middle"
  >
    <h4
      className="d-flex justify-content-center"
    >
      Latest Destinations
    </h4>
    <h5
      className="d-flex justify-content-center"
    >
       Please select Favourite Destination 
    </h5>
  </div>
</div>
`;
